[develop] Automatically retrieve instance type from capacity reservation

enrico-usai opened the pull request:
### Description of changes
Starting from now `InstanceType` is no longer a required configuration parameter.
This can be automatically retrieved from `CapacityReservationTarget/CapacityReservationId`
when it's specified at Compute Resource level.

### Tests
* Extended unit tests accordingly
* Manual test of cluster creation with and without instance type

Commit:
Call instance_types attribute rather than instance_type for SlurmComputeResource

In this way the only interface will be instance_types property,
for both SlurmComputeResource and FlexibleComputeResource (multi-instance type).

We cannot make instance_type attribute private because otherwise Marshmallow is
not able to correctly populate it.

Signed-off-by: Enrico Usai <usai@amazon.com>

Commit:
Remove useless instance_type_info property from SlurmComputeResource

There is already a `_instance_type_info` that should be used with the same scope.

Signed-off-by: Enrico Usai <usai@amazon.com>

Commit:
Automatically retrieve instance type from capacity reservation

Starting from now `InstanceType` is no longer a required configuration parameter.
This can be automatically retrieved from `CapacityReservationTarget/CapacityReservationId`
when it's specified at Compute Resource level.

Signed-off-by: Enrico Usai <usai@amazon.com>

lukeseawalker commented:
>Starting from now InstanceType is no longer a required configuration parameter.
This can be automatically retrieved from CapacityReservationTarget/CapacityReservationId
when it's specified at Compute Resource level.

what about when CapacityReservationTarget/CapacityReservationId is specified at Queue level? Is InstanceType still required?

enrico-usai commented:
> > Starting from now InstanceType is no longer a required configuration parameter.
> > This can be automatically retrieved from CapacityReservationTarget/CapacityReservationId
> > when it's specified at Compute Resource level.
> 
> what about when CapacityReservationTarget/CapacityReservationId is specified at Queue level? Is InstanceType still required?

yes, it's still required. This is a limitation of this feature. The code is able to detect the instance type from capacity reservation ONLY if it's specified at compute resource level. This because at compute resource level we have both instance-type and capacity-target info so it's easy to choose between them, but if it's not specified in the compute resource we need to have the capacity-target from the queue and this requires to pass the value to the compute resources, this is an impacting change we can avoid for now.

lukeseawalker commented on cli/src/pcluster/config/cluster_config.py:
```diff
@@ -2324,7 +2312,19 @@ def _instance_type_info(self) -> InstanceTypeInfo:
-        return self.disable_simultaneous_multithreading and self.instance_type_info.default_threads_per_core() > 1
+        return self.disable_simultaneous_multithreading and self._instance_type_info.default_threads_per_core() > 1
+
+    def _instance_id_from_capacity_target(self):
```
what about _instance_type_from_capacity_reservation?

and modify the docstring accordingly

enrico-usai replied:
yeah, I'm really not sure why I called it in this way.

enrico-usai replied:
updated and added unit tests for it.

lukeseawalker commented on cli/src/pcluster/config/cluster_config.py:
```diff
@@ -2324,7 +2312,19 @@ def _instance_type_info(self) -> InstanceTypeInfo:
+            self.capacity_reservation_target.capacity_reservation_id if self.capacity_reservation_target else None
+        )
+        if capacity_reservation_id:
+            capacity_reservations = AWSApi.instance().ec2.describe_capacity_reservations([capacity_reservation_id])
```
is this call cached?

enrico-usai replied:
Short answer: yes,

Long answer: for this call there is an internal caching mechanism (unit tested), that was already in place.

```
        missed_capacity_reservations = []
        for capacity_reservation_id in capacity_reservation_ids:
            cached_data = self.capacity_reservations_cache.get(capacity_reservation_id)
            if cached_data:
                result.append(cached_data)
            else:
                missed_capacity_reservations.append(capacity_reservation_id)
        if missed_capacity_reservations:
            response = list(
                self._paginate_results(
                    self._client.describe_capacity_reservations, CapacityReservationIds=missed_capacity_reservations
                )
            )
```
this because there are multiple `describe` calls with different capacity_reservation_ids.

lukeseawalker commented on cli/tests/pcluster/schemas/test_cluster_schema.py:
```diff
@@ -486,7 +487,33 @@ def test_scheduling_schema(mocker, config_dict, failure_message):
-            "A Compute Resource needs to specify either InstanceType or Instances.",
+            "A Compute Resource needs to specify InstanceTypes, InstanceType or CapacityReservationId.",
+        ),
+        # Failing to specify either InstanceType or Instances should return a validation error
```
should this comment moved to the next params?

lukeseawalker commented on cli/tests/pcluster/schemas/test_cluster_schema.py:
```diff
@@ -486,7 +487,33 @@ def test_scheduling_schema(mocker, config_dict, failure_message):
+            },
+            None,
+        ),
+        (
```
what if both InstanceType and CapacityReservationId are specified? Or if Instances and CapacityReservationId are specified?
Can you add a test for these conditions?

enrico-usai replied:
Added

lukeseawalker commented on cli/src/pcluster/schemas/cluster_schema.py:
```diff
@@ -1509,14 +1509,22 @@ class SlurmComputeResourceSchema(_ComputeResourceSchema):
+            )
+            if not data.get("instance_type") and not capacity_reservation_id:
+                raise ValidationError(
+                    "A Compute Resource needs to specify InstanceTypes, InstanceType or CapacityReservationId."
```
there is no InstanceTypes, but rather Instances/InstanceType

enrico-usai replied:
nice catch!

Commit:
Add unit tests for SlurmComputeResource config class

Added test for `_instance_type_from_capacity_reservation`
and for `instance_types` method.

Signed-off-by: Enrico Usai <usai@amazon.com>

lukeseawalker commented on cli/tests/pcluster/schemas/test_cluster_schema.py:
```diff
@@ -486,7 +487,33 @@ def test_scheduling_schema(mocker, config_dict, failure_message):
+                "Networking": {"SubnetIds": ["subnet-12345678"]},
+                "AllocationStrategy": "lowest-price",
+                "ComputeResources": [
+                    {"Name": "compute_resource1", "CapacityReservationTarget": {"CapacityReservationId": "id"}}
```
can you add a case when CapacityReservationTarget/CapacityReservationResourceGroupArn is specified in place of CapacityReservationTarget/CapacityReservationId ?

enrico-usai replied:
it's already there:

```
{
                "Name": "Flex-Queue",
                "Networking": {"SubnetIds": ["subnet-12345678"]},
                "AllocationStrategy": "lowest-price",
                "ComputeResources": [
                    {
                        "Name": "compute_resource1",
                        "CapacityReservationTarget": {"CapacityReservationResourceGroupArn": "arn"},
                    }
                ],
            },
            "A Compute Resource needs to specify Instances, InstanceType or CapacityReservationId.",
```
